### PR TITLE
Suggest dereferencing non-lval mutable reference on assignment

### LINKED ISSUE #46276: Poor error message when the LHS of an assignment must be dereferenced
Code (also available on [playground](https://play.rust-lang.org/?gist=1addef66d669de86966256aea3b536e4&version=nightly)):

```rust
use std::sync::Mutex;

fn main() {
    let a = Mutex::new(7);
    let mut a = a.lock().unwrap();
    a += 1;
}
```

Error:

```
error[E0368]: binary assignment operation `+=` cannot be applied to type `std::sync::MutexGuard<'_, {integer}>`
 --> src/main.rs:6:5
  |
6 |     a += 1;
  |     -^^^^^
  |     |
  |     cannot use `+=` on type `std::sync::MutexGuard<'_, {integer}>`
```

Okay, but *why* cannot `+=` be applied to `Mutexuard<'_, {integer}>`? And what can we do about it? :)

I'd prefer if the error message included something like:

```
help: consider dereferencing the left-hand side: `*a += 1`
```

Such a help message could be printed whenever the LHS of an assignment must be dereferenced. The same applies to situations in which we must dereference more than once, e.g `**a`.

### LINKED ISSUE #93980: Missing suggestion to dereference LHS of assignment when LHS is invalid
Given the following code: 

```rust
fn main() {
    let mut x: Vec<()> = vec![()];
    x.last_mut().unwrap() = ();
}
```

The current output is:

```
error[E0070]: invalid left-hand side of assignment
 --> src/main.rs:3:27
  |
3 |     x.last_mut().unwrap() = ();
  |     --------------------- ^
  |     |
  |     cannot assign to this expression

For more information about this error, try `rustc --explain E0070`.
```

Ideally the output should look like:

```
error[E0070]: invalid left-hand side of assignment
 --> src/main.rs:3:27
  |
3 |     x.last_mut().unwrap() = ();
  |     --------------------- ^
  |     |
  |     cannot assign to this expression

Note: consider dereferencing the left-hand side
3 |     *x.last_mut().unwrap() = ();
        +

For more information about this error, try `rustc --explain E0070`.
```

The root cause seems to be the coersion error being suppressed, since there is already logic that suggest dereferencing the LHS of an assignment operator. In fact, this diagnostic code path is being hit, but somewhere along the way the diagnostic is being canceled (hence the fact that we only see this "invalid left-hand side of assignment" error being emitted and not an error about not being able to relate `&mut ()` and `()`.)

### PR DESCRIPTION
**Pull request by compiler-errors**
1. Adds deref suggestions for LHS of assignment (or assign-binop) when it implements `DerefMut`
2. Fixes missing deref suggestions for LHS when it isn't a place expr

Fixes #46276
Fixes #93980

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? rust-lang/compiler

File: src/test/ui/typeck/issue-93486.stderr [hunk 1]
```diff
@@ -5,11 +5,6 @@ LL |         vec![].last_mut().unwrap() = 3_u8;
    |         -------------------------- ^
    |         |
    |         cannot assign to this expression
-   |
-help: consider dereferencing here to assign to the mutable borrowed piece of memory
-   |
-LL |         *vec![].last_mut().unwrap() = 3_u8;
-   |         +
```
Comment (wesleywiser) [inline src/test/ui/typeck/issue-93486.stderr]: Do we know why we lost this message? (It seems like it is correct unlike the one in the previous commit so I'm surprised it's missing.)

File: src/test/ui/issues/issue-5239-1.stderr [hunk 2]
```diff
@@ -5,11 +5,6 @@ LL |     let x = |ref x: isize| { x += 1; };
    |                              -^^^^^
    |                              |
    |                              cannot use `+=` on type `&isize`
-   |
-help: `+=` can be used on `isize`, you can dereference `x`
-   |
-LL |     let x = |ref x: isize| { *x += 1; };
```
Comment (compiler-errors) [inline src/test/ui/issues/issue-5239-1.stderr:11]: This is kinda a regression, but not actually... `&isize` _doesn't_ have a `+=`. Perhaps we could suggest changing the ref to `ref mut`, but that seems out of scope for this PR.

File: compiler/rustc_typeck/src/check/expr.rs [hunk 3]
```diff
@@ -1040,10 +1044,48 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        lhs_deref_ty,
+                        ty::List::empty(),
+                        self.param_env,
+                    )
+                    .may_apply();
+                if lhs_deref_ty_is_sized && self.can_coerce(rhs_ty, lhs_deref_ty) {
+                    err.span_suggestion_verbose(
+                        lhs.span.shrink_to_lo(),
+                        "consider dereferencing here to assign to the mutable \
+                        borrowed piece of memory",
```
Comment (wesleywiser) [inline compiler/rustc_typeck/src/check/expr.rs:1067]: nit: "piece of" just reads a bit awkwardly to me. Maybe "consider dereferencing here to assign to the mutably borrowed memory"?

**Review by wesleywiser — APPROVED** [review]
r=me with or without nit. Nice work! 💯

File: compiler/rustc_typeck/src/check/demand.rs [hunk 4]
```diff
@@ -696,28 +696,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-                                        "consider dereferencing here to assign to the mutable \
-                                         borrowed piece of memory"
-                                            .to_string(),
-                                        "*".to_string(),
-                                        Applicability::MachineApplicable,
-                                        true,
-                                    ));
-                                }
+                                // Suppressing this diagnostic, we'll properly print it in `check_expr_assign`
+                                return None;
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/demand.rs:705]: For future reference, adding a `delay_span_bug` here might be a good idea to protect from miscompilations in the face of refactors.

Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/demand.rs:705]: shoot, yeah. i can add one in a follow-up.

Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/demand.rs:705]: Actually, this just suppresses a suggestion, not the error itself. And the other case I actually did suppress a real error, I downgraded it to a delayed bug. See: https://github.com/rust-lang/rust/pull/94639/files#diff-c46757032f463a1170b40e5c41d569ce058668cfe259774fb8a7936e3fdd82f4R57

Comment (estebank) [reply compiler/rustc_typeck/src/check/demand.rs:705]: @compiler-errors perfect!
